action/protocol/poll/util.go

```diff
@@ -170,17 +176,78 @@ func setCandidates(
 			zap.String("score", candidate.Votes.String()),
 		)
 	}
-	_, err := sm.PutState(&candidates, protocol.LegacyKeyOption(candidatesutil.ConstructKey(height)))
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if hu.IsPre(config.Easter, height) {
+		_, err := sm.PutState(&candidates, protocol.LegacyKeyOption(candidatesutil.ConstructLegacyKey(height)))
+		return err
+	}
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.PutState(&candidates, protocol.KeyOption(nextKey[:]))
 	return err
 }
 
-// setKickoutBlackList sets the blacklist for kick-out for corresponding epoch
+// setKickoutBlackList sets the blacklist for kick-out with next key
 func setKickoutBlackList(
 	sm protocol.StateManager,
 	blackList *vote.Blacklist,
-	epochNum uint64,
 ) error {
-	blackListKey := candidatesutil.ConstructBlackListKey(epochNum)
-	_, err := sm.PutState(blackList, protocol.LegacyKeyOption(blackListKey))
+	blackListKey := candidatesutil.ConstructKey(candidatesutil.NxtKickoutKey)
+	_, err := sm.PutState(blackList, protocol.KeyOption(blackListKey[:]))
 	return err
 }
+
+// setUnproductiveDelegates sets the upd struct with updkey
+func setUnproductiveDelegates(
+	sm protocol.StateManager,
+	upd *vote.UnproductiveDelegate,
+) error {
+	updKey := candidatesutil.ConstructKey(candidatesutil.UnproductiveDelegateKey)
+	_, err := sm.PutState(upd, protocol.KeyOption(updKey[:]))
+	return err
+}
+
+// shiftCandidates updates current data with next data of candidate list
+func shiftCandidates(sm protocol.StateManager) error {
+	zap.L().Debug("Shift candidatelist from next key to current key")
+	var next state.CandidateList
+	nextKey := candidatesutil.ConstructKey(candidatesutil.NxtCandidateKey)
+	_, err := sm.State(&next, protocol.KeyOption(nextKey[:]))
+	if err != nil {
+		return errors.Wrap(
+			err,
+			"failed to read next blacklist when shifting to current blacklist",
+		)
+	}
+	curKey := candidatesutil.ConstructKey(candidatesutil.CurCandidateKey)
+	_, err = sm.PutState(&next, protocol.KeyOption(curKey[:]))
```

Comment:
do a redundant checking that the height returned by sm.State and sm.PutState are identical